src/connection.ts

```diff
@@ -4,70 +4,235 @@ import { AbortOptions } from '@libp2p/interfaces';
+  }
+
+  private countStream(protocol: string, direction: ic.Direction): number {
+    return this.streams.filter((s) => s.stat.protocol === protocol && s.stat.direction === direction).length;
+  }
+
+  async newStream(protocols: string | string[], options: AbortOptions = {}): Promise<ic.Stream> {
+    let label = genUuid().slice(0, 8);
+    let openPromise = defer();
+    let abortedPromise = defer();
+    let controller: TimeoutController | undefined;
+    let metrics = this.components.getMetrics();
+    let openError: Error | undefined;
+
+    log.trace(`opening new stream with protocols: ${protocols}`);
+
+    // timeout in case no abort options are provided
+    if (options.signal == null) {
+      log.trace(`[stream: ${label}] no abort signal provided, creating timeout controller`);
+      controller = new TimeoutController(OPEN_STREAM_TIMEOUT);
+      options.signal = controller.signal;
+    }
+
+    options.signal.onabort = () => {
+      log.trace(`[stream: ${label}] abort called - ${options.signal?.reason}`);
+      openError = new Error(options.signal?.reason || 'aborted');
+      abortedPromise.resolve();
+    };
+
+    log.trace(`[stream: ${label}] peerconnection state: ${this.peerConnection.connectionState}`);
+    let channel = this.peerConnection.createDataChannel(label);
+    channel.onopen = (_evt) => {
+      log.trace(`[stream: ${label}] data channel opened`);
+      openPromise.resolve();
+    };
+    channel.onerror = (_evt) => {
+      log.trace(`[stream: ${label}] data channel error: ${(_evt as RTCErrorEvent).error}`);
+      openError = new Error(`data channel error`);
+      abortedPromise.resolve();
+    };
+
+    log.trace(`[stream: ${label}] datachannel state: ${channel.readyState}`);
+    await Promise.race([openPromise.promise, abortedPromise.promise]);
+
+    // check for error
+    if (openError) {
+      // TODO: Better errors
+      throw openError;
+    }
+
+    let rawStream = new WebRTCStream({
+      channel,
+      stat: {
+        direction: 'outbound',
+        timeline: {
+          open: new Date().getTime(),
+        },
+      },
+    });
+
+    let { stream, protocol } = await msselect(rawStream, protocols, { signal: options.signal });
+    log.trace(`[stream ${label}] select protocol - ${protocol}`);
+    // check if stream is within limit after protocol has been negotiated
+    rawStream.stat.protocol = protocol;
+    let result = this.wrapMsStream(rawStream, stream);
+    // check if stream can be accomodated
+    if (metrics) {
+      metrics.trackStream({ stream, protocol, remotePeer: this.remotePeer });
+    }
+
+    this.addStream(result);
+    return result;
   }
 
   addStream(stream: ic.Stream): void {
-    throw new Error('not implemented');
+    let protocol = stream.stat.protocol!;
+    let direction = stream.stat.direction;
+    if (this.countStream(protocol, direction) === this.findStreamLimit(protocol, direction)) {
```

Comment:
Although I  understand it would be an incredibly bizarre implementation that did this, it seems to me to be at least _possible_ that the data being fetched by findStreamLimit _could_ theoretically change at runtime. So I'd be inclined to test with something like `>=`